src/abstractdataframe/join.jl

```diff
@@ -337,10 +356,28 @@ function Base.join(df1::AbstractDataFrame,
                 push!(leftonly_ixs, l_ix)
             end
         end
-        return joiner.dfl[leftonly_ixs, :]
+        joined = joiner.dfl[leftonly_ixs, :]
     else
         throw(ArgumentError("Unknown kind of join requested: $kind"))
     end
+
+    if indicator !== nothing
+        left = joined[Symbol(indicator_cols[1])]
+        left = Int.(coalesce.(left, false))
```

Comment:
This allocates two copies of the column, and introduces a type instability by changing the type of `left`. You don't actually need to do this here: just call `coalesce(left[i], false)` inside the comprehension (`Int` isn't needed if you speficy `UInt8[...]` before the comprehension).

Reply:
OK, but I don't think that's a type instability (or what is generally meant by "type instability" in Julia). Isn't a "type instability" a situation in which the type of a variable depends on the values of data. When that happens, it causes problems for the compiler because the compiler is only conditioning its code on function input types, not input values, so it can't make a clean type inference. e.g.

```
function x(a::Integer)
    if a % 2  == 0
        return a
    else
       return "a is odd"
end
```
Is type unstable because if a is even , the function returns an integer, if its odd it return a string. And the compiler doesn't condition on the *value* of a, just it's type (here, `Integer`). 

Here, by contrast, the type of the vector to which `left` is bound changes, but in an entirely transparent way. The compiler should have no problems with this. 

Also, I think you may be underestimating the compiler -- it knows not to double allocate: 

```
julia> function test_basic(left)
           return UInt8.(coalesce.(left, false))
           end
julia> function test_fancy(left)
           return UInt8[coalesce(left[i], false) for i in 1:length(left)]
           end
julia> left = rand([true, missing], 10000)

julia> @btime test_fancy(left)
  77.040 μs (1 allocation: 9.94 KiB)

julia> @btime test_basic(left)
  80.942 μs (1 allocation: 9.94 KiB)
```

Reply:
You are over-estimating the compiler here I guess. Here is a short example that this is type unstable:

```
julia> using DataFrames

julia> x = DataFrame(rand(3,4))
3×4 DataFrames.DataFrame
│ Row │ x1       │ x2       │ x3       │ x4       │
├─────┼──────────┼──────────┼──────────┼──────────┤
│ 1   │ 0.825846 │ 0.719652 │ 0.968676 │ 0.892195 │
│ 2   │ 0.46385  │ 0.51642  │ 0.229933 │ 0.675513 │
│ 3   │ 0.790047 │ 0.811756 │ 0.635098 │ 0.836908 │

julia> function f(df)
       left = df[1]
       left = floor.(Int, left)
       end
f (generic function with 2 methods)

julia> f(x)
3-element Array{Int64,1}:
 0
 0
 0

julia> @code_warntype f(x)
Variables:
  #self# <optimized out>
  df::DataFrames.DataFrame
  left::Any
  selected_column <optimized out>

Body:
  begin
      left::Any = (Base.arrayref)((Core.getfield)(df::DataFrames.DataFrame, :columns)::Array{T,1} where T, 1)::Any # line 3:
      SSAValue(0) = (Base.broadcast)(Main.floor, Main.Int, left::Any)::Any
      left::Any = SSAValue(0)
      return SSAValue(0)
  end::Any
```

Sometimes compiler can optimize out such code, but it is not guaranteed.

In general type-stability is a situation when type of a value bound to a variable is unpredictable or may change (even if predictably) I would say.

Reply:
FYI, re: https://discourse.julialang.org/t/why-is-changing-type-of-variable-problematic/11722/17 : Looks like this is getting much better in 0.7. 

Re your example @bkamins , I think that still has problems in 0.7, but that's because of the general issues with type inference and `DataFrames`. Doing the same with regular vectors seems fine:

```
function f2(m)
              onecol = m[:,1]
              onecol = floor.(Int, onecol)
end

julia> m = rand(3,4)
3×4 Array{Float64,2}:
 0.841133  0.613176  0.145906   0.526924
 0.388555  0.709457  0.0679149  0.633461
 0.858938  0.835699  0.599265   0.814602

julia> f2(m)
3-element Array{Int64,1}:
 0
 0
 0

julia> @code_warntype f2(m)
Body::Array{Int64,1}
2 1 ── %1   = Base.arraysize(%%m, 1)::Int64 getindex
  │    %2   = Base.arraysize(%%m, 2)::Int64  to_indices
  │    %3   = Base.slt_int(%1, 0)::Bool       axes
  │    %4   = Base.ifelse(%3, 0, %1)::Int64    map
  │    %5   = new(Base.OneTo{Int64}, %4)::Base.OneTo{Int64}
  │    %6   = Base.slt_int(%2, 0)::Bool          Type
  │           Base.ifelse(%6, 0, %2)              max
  │    %8   = new(Base.Slice{Base.OneTo{Int64}}, %5)::Base.Slice{Base.OneTo{Int64}}
  └───        goto 6 if not true             _getindex
  2 ── %10  = Core.tuple(%8, 1)::Tuple{Base.Slice{B

[lots more but no warnings...]
```

Reply:
Good to know. However in many situations it's clearer to avoid using the same name for values of different types.